connectors/salesforce/src/main/java/com/vmware/connectors/salesforce/SalesforceController.java

```diff
@@ -227,16 +238,127 @@ private URI makeUri(
+
         } else {
             // Contact doesn't exist in salesforce. Return a card to show accounts that are related to sender domain.
             logger.debug("Returning accounts info for domain: {} ", senderDomain);
             return makeCardsFromSenderDomain(auth, baseUrl, routingPrefix, userEmail, senderEmail, senderDomain, locale);
         }
     }
 
+    // Contact details card and Opportunity cards.
+    private Flux<Card> getCardsForSender(JsonDocument contactDetails,
+                                         JsonDocument oppIds,
+                                         String baseUrl,
+                                         String auth,
+                                         String routingPrefix,
+                                         Locale locale,
+                                         HttpServletRequest request) {
+
+        Flux<Card> userDetailCard = Flux.just(createUserDetailsCard(contactDetails, routingPrefix, locale, request));
+
+        int count = oppIds.read("$.totalSize");
+        if (count > 0) {
+
+            List<String> Ids = oppIds.read("$.records[*].Opportunity.Id");
+
+            Flux<Card> opportunityCards = retrieveOpportunities(Ids, baseUrl, auth)
+                    .flatMapMany(document -> createOpportunityCards(document, routingPrefix, locale, request));
+
+            return Flux.concat(userDetailCard, opportunityCards);
+        }
+
+        return userDetailCard;
+    }
+
+    private Mono<JsonDocument> retrieveOppIds(String senderEmail,
+                                              String baseUrl,
+                                              String auth) {
+        String soql = String.format(QUERY_FMT_CONTACT_OPPORTUNITY, senderEmail);
+        return rest.get()
+                .uri(makeSoqlQueryUri(baseUrl, soql))
+                .header(AUTHORIZATION, auth)
+                .retrieve()
+                .bodyToMono(JsonDocument.class);
+    }
+
+    private Mono<JsonDocument> retrieveOpportunities(List<String> oppIds, String baseUrl, String auth) {
+        String idsFormat = oppIds.stream().collect(Collectors.joining("', '"));
+
+        String soql = String.format(QUERY_FMT_OPPORTUNITY_INFO, idsFormat);
+        return rest.get()
+                .uri(makeSoqlQueryUri(baseUrl, soql))
+                .header(AUTHORIZATION, auth)
+                .retrieve()
+                .bodyToMono(JsonDocument.class);
+    }
+
+    private Flux<Card> createOpportunityCards(JsonDocument opportunities,
+                                              String routingPrefix,
+                                              Locale locale,
+                                              HttpServletRequest request) {
+
+        int oppCount = opportunities.read("$.totalSize");
+
+        List<Card> oppCards = new ArrayList<>();
+        for (int oppIndex = 0; oppIndex < oppCount; oppIndex++) {
+
+            String name = opportunities.read(String.format("$.records[%d].Name", oppIndex));
```

Comment:
I'm assuming the generics and casting gets ugly if you try to loop `Map<String, Object>`, but maybe try to reduce the duplicate string format parts with something like:

```java
for (int oppIndex = 0; oppIndex < oppCount; oppIndex++) {

    String prefix = String.format("$.records[%d]", oppIndex);

    String name = opps.read(prefix + ".Name");
    List<Object> feedComments = opps.read(prefix + ".Feeds.records[*]");

    CardBody.Builder cardBodyBuilder = new CardBody.Builder()
            .setDescription(cardTextAccessor.getMessage("opportunity.description", locale))
            .addField(bodyField("opportunity.account", opps.read(prefix + ".Account.Name"), locale))
            .addField(bodyField("opportunity.account.owner", opps.read(prefix + ".Account.Owner.Name"), locale))
            .addField(bodyField("opportunity.closedate", opps.read(prefix + ".CloseDate"), locale))
            .addField(bodyField("opportunity.stage", opps.read(prefix + ".StageName"), locale))
            .addField(bodyField("opportunity.amount", opps.read(prefix + ".Amount"), locale));
```

(Note: I shortened opportunities and generalBodyField so things fit better on 1 line)